R/methods_formula.R

```diff
@@ -0,0 +1,204 @@
+
+  if ('projection' %in% class(object) ||
+      (length(object)>0 && 'projection' %in% class(object[[1]]))) {
+    proj <- object
+  } else {
+    # reference model or varsel object obtained, so run the projection
+    proj <- project_poc(object = object, nv = nv, ...)
+  }
+
+  if (!.is_proj_list(proj)) {
+    proj <- list(proj)
+  } else {
+    # proj is not a projection object
+    if(any(sapply(proj, function(x) !('family_kl' %in% names(x)))))
+      stop(paste('proj_linpred only works with objects returned by',
+                 ' varsel, cv_varsel or project'))
+  }
+
+  ## FIXME: proj$sub_fit always exists, but does $coefficients?
+  projected_sizes <- sapply(proj, function(x) NCOL(x$sub_fit$coefficients))
+  nv <- list(...)$nv %ORifNULL% projected_sizes
+
+  if (!all(nv %in% projected_sizes))
+    stop(paste0('Linear prediction requested for nv = ',
+                paste(nv, collapse = ', '),
+                ', but projection performed only for nv = ',
+                paste(projected_sizes, collapse = ', '), '.'))
+
+  projs <- Filter(function(x) NCOL(x$sub_fit$coefficients) %in% nv, proj)
+  names(projs) <- nv
+
+  xnew_df <- is.data.frame(xnew)
+  if (xnew_df) {
+    terms <- unique(unlist(lapply(projs, function(x) unlist(unname(x$vind)))))
+    xnew <- .df_to_model_mat(xnew, terms)
+  }
+
+  if (!is.matrix(xnew))
+    stop('xnew not provided in the correct format. See ?proj-pred.')
+
+  vind <- list(...)$vind
+  if (!is.null(vind) && NCOL(xnew) != length(vind))
+    stop(paste('The number of columns in xnew does not match with the given',
+               'number of variable indices (vind).'))
+
+  # set random seed but ensure the old RNG state is restored on exit
+  rng_state_old <- rngtools::RNGseed()
+  on.exit(rngtools::RNGseed(rng_state_old))
+  set.seed(seed_samp)
+
+  preds <- lapply(projs, function(proj) {
+    if (xnew_df) {
+      xtemp <- xnew[, min(1, length(proj$vind)):length(proj$vind), drop = F]
+    } else if (!is.null(vind)) {
+      # columns of xnew are assumed to match to the given variable indices
+      xtemp <- xnew
+    } else {
+      # fetch the right columns from the feature matrix
+      xtemp <- xnew[, unique(unname(unlist(proj$vind))), drop = F]
+    }
+    mu <- proj$family_kl$mu_fun(proj$sub_fit, xnew=xtemp)
+
+    fun(proj, mu, offsetnew, weightsnew)
+  })
+
+  .unlist_proj(preds)
+}
+
+#' @rdname proj-pred
+#' @export
+proj_linpred_poc <- function(object, xnew, ynew = NULL, offsetnew = NULL,
+                             weightsnew = NULL, nv = NULL, transform = FALSE,
+                             integrated = FALSE, ...) {
+
+  # function to perform to each projected submodel
+  fun <- function(proj, mu, offset, weights) {
+    pred <- t(mu)
+    if (!transform) pred <- proj$family_kl$linkfun(pred)
+    if (integrated) {
+      # average over the parameters
+      pred <- as.vector( proj$weights %*% pred )
+    } else if (!is.null(dim(pred)) && dim(pred)[1]==1) {
```

Comment:
I guess what we are saying is that if the user does not want the integrated behaviour but only has one prediction, work on it as a vector. Otherwise work with it as a matrix. Maybe the matrix behaviour holds in this case as well, I'll look it up.